File: android/src/main/java/fi/mankelisolutions/display_brightness/DisplayBrightness.kt
package fi.mankelisolutions.display_brightness

import android.app.Activity
import android.content.ContentResolver
import android.database.ContentObserver
import android.net.Uri
import android.os.Handler
import android.os.Looper
import android.provider.Settings
import androidx.annotation.Keep

/**
 * Controls app-level screen brightness and monitors system brightness changes.
 *
 * Uses [android.view.WindowManager.LayoutParams.screenBrightness] to control
 * the app window brightness (no permissions required).
 *
 * @param activity the current Activity used to access the window and content resolver.
 */
@Keep
class DisplayBrightness(private val activity: Activity) {

    private val contentResolver: ContentResolver = activity.contentResolver
    private val brightnessUri: Uri =
        Settings.System.getUriFor(Settings.System.SCREEN_BRIGHTNESS)
    private var contentObserver: ContentObserver? = null
    private var callback: BrightnessCallback? = null

    /**
     * Returns the current effective screen brightness (0.0–1.0).
     *
     * If the app has set a custom brightness via [setBrightness], returns that value.
     * Otherwise, reads the system brightness setting and normalizes it from 0–255 to 0.0–1.0.
     * Returns null if the system brightness cannot be read.
     */
    @get:Keep
    val brightness: Double?
        get() {
            val layoutBrightness = activity.window.attributes.screenBrightness
            val raw = if (layoutBrightness < 0) {
                // App hasn't overridden brightness — read system setting
                try {
                    Settings.System.getInt(
                        contentResolver,
                        Settings.System.SCREEN_BRIGHTNESS
                    ) / 255.0
                } catch (e: Settings.SettingNotFoundException) {
                    null
                }
            } else {
                layoutBrightness.toDouble()
            }
            return if (raw != null) Math.round(raw * 100.0) / 100.0 else null
        }

    /**
     * Sets the app-level screen brightness.
     *
     * @param brightness value between 0.0 (darkest) and 1.0 (brightest).
     */
    @Keep
    fun setBrightness(brightness: Double) {
        val layoutParams = activity.window.attributes
        layoutParams.screenBrightness = brightness.toFloat()
        activity.window.attributes = layoutParams
    }

    /**
     * Starts observing system screen brightness changes.
     *
     * The [callback] will be invoked on the main thread whenever the system
     * brightness setting changes, receiving the effective brightness value (0.0–1.0).
     *
     * Only one callback can be active at a time. Calling this method again
     * replaces the previous callback (and re-registers the observer).
     */
    @Keep
    fun startObserving(callback: BrightnessCallback) {
        stopObserving()

        this.callback = callback

        contentObserver = object : ContentObserver(Handler(Looper.getMainLooper())) {
            override fun onChange(selfChange: Boolean) {
                super.onChange(selfChange)
                val systemBrightness = try {
                    Settings.System.getInt(
                        contentResolver,
                        Settings.System.SCREEN_BRIGHTNESS
                    ) / 255.0
                } catch (e: Settings.SettingNotFoundException) {
                    0.5
                }

                // If app currently overrides brightness, clear it to let the system brightness take effect
                val layoutParams = activity.window.attributes
                if (layoutParams.screenBrightness >= 0) {
                    layoutParams.screenBrightness = -1.0f
                    activity.window.attributes = layoutParams
                }

                val roundedBrightness = Math.round(systemBrightness * 100.0) / 100.0
                this@DisplayBrightness.callback?.onBrightnessChanged(roundedBrightness)
            }
        }

        contentResolver.registerContentObserver(brightnessUri, false, contentObserver!!)
    }

    /**
     * Stops observing screen brightness changes and removes the callback.
     */
    @Keep
    fun stopObserving() {
        contentObserver?.let {
            contentResolver.unregisterContentObserver(it)
        }
        contentObserver = null
        callback = null
    }
}

File: ios/display_brightness/Sources/display_brightness/DisplayBrightness.swift
import UIKit

/// Callback protocol for screen brightness changes.
@objc public protocol BrightnessCallback {
    /// Called when the screen brightness changes.
    ///
    /// - Parameter brightness: the new brightness value (0.0–1.0).
    @objc func onBrightnessChanged(_ brightness: Double)
}

/// Controls app-level screen brightness and monitors brightness changes.
///
/// Uses `UIScreen.main.brightness` to control and read screen brightness.
/// No permissions are required.
@objc public class DisplayBrightness: NSObject {
    private var observer: NSObjectProtocol?
    private var callback: BrightnessCallback?

    @objc public override init() {
        super.init()
    }

    /// Returns the current screen brightness value (0.0–1.0).
    @objc public var brightness: Double {
        return (Double(UIScreen.main.brightness) * 100.0).rounded() / 100.0
    }

    /// Sets the screen brightness.
    ///
    /// - Parameter brightness: value between 0.0 (darkest) and 1.0 (brightest).
    @objc public func setBrightness(_ brightness: Double) {
        UIScreen.main.brightness = CGFloat(brightness)
    }

    /// Starts observing screen brightness changes.
    ///
    /// The callback will be invoked whenever the screen brightness changes,
    /// receiving the new brightness value (0.0–1.0).
    ///
    /// Only one callback can be active at a time. Calling this method again
    /// replaces the previous callback.
    @objc public func startObserving(callback: BrightnessCallback) {
        stopObserving()

        self.callback = callback

        observer = NotificationCenter.default.addObserver(
            forName: UIScreen.brightnessDidChangeNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            guard let self = self else { return }
            self.callback?.onBrightnessChanged(self.brightness)
        }
    }

    /// Stops observing screen brightness changes and removes the callback.
    @objc public func stopObserving() {
        if let observer = observer {
            NotificationCenter.default.removeObserver(observer)
        }
        observer = nil
        callback = nil
    }

    deinit {
        stopObserving()
    }
}
